### PR TITLE
Gherkin catalog

### PR DESCRIPTION
**Pull request by Jaykul**
Shipping gherkin.dll (and Newtonsoft.Json) for both .NET35 (PS2) and Core

Note that the .NET 3.5 build (in the "legacy" folder) works fine on .NET 4.x so I didn't bother shipping the .NET 4.x specific version -- although we could, if we want to go for the XXL module size ;-)

Also, if you want to build your own .NET 3.5 version of gherkin.dll, you need [my fork](https://github.com/jaykul/gherkin-dotnet/tree/Pester-PowerShell2-Net35)...

I added a catalog file for the assemblies (use `Test-FileCatalog catalog.cat`), but I noticed something weird that I cannot explain. The hashes calculated by `New-FileCatalog` and `Test-FileCatalog` are different than the ones calculated by `Get-FileHash`...

```
I ♥ PS> ls *\*.dll -Recurse | Get-FileHash | ft @{N="Path";E={Resolve-Path $_.Path -relative}}, Hash, Algorithm

Path                         Hash                                                             Algorithm
----                         ----                                                             ---------
.\core\Gherkin.dll           352A6A598BA8D1E350F497791A9FB9AC64BA11E977FF6CCB7432D8516DDF7685 SHA256
.\core\Newtonsoft.Json.dll   5D96EE51B2AFF592039EEBC2ED203D9F55FDDF9C0882FB34D3F0E078374954A5 SHA256
.\legacy\Gherkin.dll         2BF80E108FC1A112D9AC9E631D29903028EEBABD2A7831ACFCE88CCBEFDF773E SHA256
.\legacy\Newtonsoft.Json.dll EFD8CF9A3BC2AB4E15FA33D42771E18D78539759CBF30652DF4C43E6825CE5F0 SHA256


I ♥ PS> Test-FileCatalog .\catalog.cat -Detailed | % PathItems

Key                        Value
---                        -----
core\Gherkin.dll           D8C0312ECB4D12899F13AD9D6185D15EA42BA283A35044E19302052FDA02BAFC
core\Newtonsoft.Json.dll   AC04BBF9651A748685CD77D9B13D5A8BA0B9CC67E078185635685547D442CBA5
legacy\Gherkin.dll         7941A923F79AEABDFB45F544BE6159376260A81B6AC1BF18A5F868B402C20650
legacy\Newtonsoft.Json.dll F5973B9F2BA68F541C6F840E3867D7D251E1B299F6147BE3FC1DBAC167695462
```

### CODE REVIEW

**Comment by JPRuskin** [comment]
I might be missing some information about what -CatalogVersion does, but it looks like New-FileCatalog uses SHA1 as a default, whilst SHA256 was used by Get-FileHash above.

    PS C:\..\>New-FileCatalog -CatalogFilePath .\
    
    Mode                LastWriteTime         Length Name
    ----                -------------         ------ ----
    -a----       07/09/2017     10:08           2464 catalog.cat

    PS C:\..\>Test-FileCatalog .\catalog.cat -Detailed
    
    Status        : Valid
    HashAlgorithm : SHA1
    CatalogItems  : <...>
    PathItems     : <...>

This may not be the case with your catalog?

**Comment by it-praktyk** [comment]
I can't explain it but I see differences in sums for all dll files what you added - after merging with my branches.

I tried few times - two times on Windows (on PowerShell  5.1.15063.502  and in a pure git console for Windows - mintty 2.7.7 x86_64-pc-msys) and on Linux. Yesterday I suspected that something is wrong because I had segmentation fault of git (in PowerShell when I've tried cherry-pick your commits on top of my DevV4-1_Linux_Compatibility_2.

I did it using different methods - means
1. cherry-pick of separate commits
2. merge your branch to mine

The results what I've

```
PS >ls *\*.dll -Recurse | Get-FileHash | ft @{N="Path";E={Resolve-Path $_.Path -relative}}, Hash, Algorithm

Path                             Hash                                                             Algorithm
----                             ----                                                             ---------
.\lib\core\Gherkin.dll           96CDBB398AC30D3B95D1D13493917F3B9FEB77B0B52C2DF39FD9F719DFF63AE4 SHA256
.\lib\core\Newtonsoft.Json.dll   5AA7325DC4167A8A6E1C9033AA7C128E0D82570069175EEAFB001763A53FA1D1 SHA256
.\lib\legacy\gherkin.dll         59D59791DC7B7217E54F323665B7E5AC5789A77D8B51D21E9C4F66DFF88C20A7 SHA256
.\lib\legacy\Newtonsoft.Json.dll 25C2A03C49FE1823617B0D18CBB51FE287C7BF8DE1F54833699EB9335C0DD54D SHA256
```

Can someone of you try to repeat it on your computer?

File: Functions/Gherkin.ps1 [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Work around bug in PowerShell 2 type loading...
-Microsoft.PowerShell.Core\Import-Module -Name "${Script:PesterRoot}\lib\Gherkin.dll"
+if(!$IsCoreClr) {
+    Microsoft.PowerShell.Core\Import-Module -Name "${Script:PesterRoot}\lib\core\Gherkin.dll"
+} else {
+    Microsoft.PowerShell.Core\Import-Module -Name "${Script:PesterRoot}\lib\legacy\Gherkin.dll"
```
Comment (it-praktyk) [inline Functions/Gherkin.ps1]: The file `lib\legacy\gherkin.dll` has the first letter small and I don't like that the first letter for `lib\core\Gherkin.dll` is not small too.

Additionally, to achieve compatibility with Linux and macOS we need use code like below
```
if(!$IsCoreCLR) {
    Microsoft.PowerShell.Core\Import-Module -Name "${Script:PesterRoot}\lib\legacy\gherkin.dll"
} else {
    [String]$GherkinCoreDllPath = "${Script:PesterRoot}{0}lib{0}core{0}Gherkin.dll" -f [System.IO.Path]::DirectorySeparatorChar
    Microsoft.PowerShell.Core\Import-Module -Name $GherkinCoreDllPath
}
```

**Comment by it-praktyk** [comment]
My issue with binary files is probably caused by [my "smart" commit](https://github.com/it-praktyk/Pester/commit/ba38cc6fdbcc130434b4628fe8865c3c065e8308).

**Comment by it-praktyk** [comment]
I can confirm now. Mentioned previously "smart" commit caused that end of lines - even in binary files - were replaced from LF to CRLF.

**Comment by Jaykul** [comment]
Whew! Glad you figured that out, I left for a long weekend and I was thinking: how am I going to figure that out on my android phone!? ;-)

See, this is why LF only, FTW

**Comment by it-praktyk** [comment]
@Jaykul, now - when results of your work are part of [a feature branch](https://github.com/pester/Pester/tree/Dev4-1_PSCore_compatibility) - can we close the current pull request?

**Comment by Jaykul** [comment]
Sure.

I'm still a bit confused why anyone made a big deal about needing core compat if Pester wasn't working there anyway ;-)

**Comment by it-praktyk** [comment]
Thx

**Pull request closed without merging**